src/OrchardCore.Modules/OrchardCore.ContentPreview/Views/Preview/Index.cshtml

```diff
@@ -34,91 +34,112 @@
-    var iframe = document.getElementById('iframe');
-    var notConnectedWarning = document.getElementById('notConnectedWarning');
+    iframe.onload = function () {
+        previewHeadRefresh = true;
+    }
 
     var scrollTop;
     var rendering;
 
     function renderPreview(value) {
 
-        renderRequested = true;
-
-        // Squashes all event calls into one
         if (rendering) {
+            clearTimeout(previewRenderTimer);
+            previewRenderTimer = setTimeout(function () { renderPreview(value); }, 150);
             return;
         }
 
-        // Pump renderPreview calls
-        while (renderRequested) {
-            renderRequested = false;
-            rendering = true;
-            try {
+        rendering = true;
+        clearTimeout(previewRenderTimer);
 
-                var formData = JSON.parse(value);
-                if (!formData) return;
+        try {
+            var formData = JSON.parse(value);
+            if (!formData) {
+                rendering = false;
+                return;
+            }
 
-                scrollTop = $(iframe.contentWindow).scrollTop();
+            scrollTop = $(iframe.contentWindow).scrollTop();
 
-                $.post(renderPreviewUrl, formData)
-                    .done(function (data) {
-                        if (!iframe || !iframe.contentWindow) {
-                            iframe.contentWindow.document.open();
-                            iframe.contentWindow.document.close();
+            $.post(renderPreviewUrl, formData)
+                .done(function (data) {
+                    if (iframe && iframe.contentWindow) {
+                        if (iframe.contentWindow.document.head.innerHTML == '') {
+                            iframe.contentWindow.document.write(data);
+                            previewHeadRefresh = false;
                         }
-
-                        iframe.contentWindow.document.body.innerHTML = '';
-                        iframe.contentWindow.document.write(data);
-                        $(iframe.contentWindow).scrollTop(scrollTop);
-                        $(serverErrorWarning).hide();
-                    })
-                    .fail(function (data) {
-                        if (data.responseJSON && data.responseJSON.errors) {
-                            var ul = $('#serverErrorWarning ul').empty();
-                            data.responseJSON.errors.forEach(function (error) {
-                                console.error(error);
-                                $('<li/>').text(error).appendTo(ul);
-                            });
-                            $(serverErrorWarning).show();
+                        else {
+                            if (previewHeadRefresh) {
+                                previewHeadRefresh = false;
+                                var head = data.substring(data.indexOf('<head>') + 6, data.indexOf('</head>'));
```

Comment:
Explain?

Reply:
Just posted a last comment to explain things, here the related points.

- Synchronizes head data e.g after having navigated through the preview windows.

- And prevents from injecting head data under the body tag. E.g, while previewing, the collapsable menu was not always working.

- Note: Normally for the frame the right way would be to use document open(), write() and then close(), but this re-add some flickering and it's more difficult to handle the scrolling position.